## Практическое задание

<ol>
    <li>Попробуйте обучить нейронную сеть на Keras с другими параметрами. 
        Опишите в комментарии к уроку - какой результата вы добились от нейросети? Что помогло вам улучшить ее точность?</li>
    <li>Поработайте с документацией Keras.</li>
</ol>

In [169]:
# The full neural network code!
###############################
import numpy as np
import mnist
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical

def accuracy(y_true, y_pred):
    return np.mean(y_true == y_pred)

import warnings
warnings.filterwarnings('ignore')

In [170]:
# Seed value
# Apparently you may use different seed values at each stage
seed_value= 42

# 1. Set `PYTHONHASHSEED` environment variable at a fixed value
import os
os.environ['PYTHONHASHSEED']=str(seed_value)

# 2. Set `python` built-in pseudo-random generator at a fixed value
import random
random.seed(seed_value)

# 3. Set `numpy` pseudo-random generator at a fixed value
import numpy as np
np.random.seed(seed_value)

# 4. Set the `tensorflow` pseudo-random generator at a fixed value
import tensorflow as tf
tf.random.set_seed(seed_value)

In [181]:
train_images = mnist.train_images()
train_labels = mnist.train_labels()
test_images = mnist.test_images()
test_labels = mnist.test_labels()

# Normalize the images.
train_images = (train_images / 255) - 0.5
test_images = (test_images / 255) - 0.5


# Flatten the images.
train_images = train_images.reshape((-1, 784))
test_images = test_images.reshape((-1, 784))

In [182]:
# Build the model.
model = Sequential([
  Dense(64, activation='elu', input_shape=(784,)),
  Dense(64, activation='elu'),
    Dense(64, activation='elu'),
#         Dense(64, activation='relu'),
  Dense(10, activation='softmax'),
])

In [175]:
# Build the model.
model = Sequential([
  Dense(64, activation='relu', input_shape=(784,)),
  Dense(128, activation='relu'),
    Dense(64, activation='relu'),
  Dense(10, activation='softmax'),
])

In [161]:

# Compile the model.
model.compile(
  optimizer='Adadelta',
  loss='cosine_similarity',
  metrics=['top_k_categorical_accuracy'],
)

In [176]:

# Compile the model.
model.compile(
  optimizer='Adam',
  loss='cosine_similarity',
  metrics=['top_k_categorical_accuracy'],
)

In [180]:


# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=5,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

# Save the model to disk.
# model.save_weights('model1.h5')

# Load the model from disk later using:
# model.load_weights('model.h5')


# Predict on the first 10 test images.
predictions = model.predict(test_images[:100])

# Print our model's predictions.
print(np.argmax(predictions[:10], axis=1)) # [7, 2, 1, 0, 4]

# Check our predictions against the ground truths.
print(test_labels[:10]) # [7, 2, 1, 0, 4]

print(f'Accuracy: {accuracy(test_labels[:100], np.argmax(predictions, axis=1))}')

Epoch 1/5
60000/60000 [==============================] - 4s 60us/step - loss: -0.9742 - top_k_categorical_accuracy: 0.9988
Epoch 2/5
60000/60000 [==============================] - 3s 51us/step - loss: -0.9750 - top_k_categorical_accuracy: 0.9988
Epoch 3/5
60000/60000 [==============================] - 3s 53us/step - loss: -0.9746 - top_k_categorical_accuracy: 0.9986
Epoch 4/5
60000/60000 [==============================] - 3s 53us/step - loss: -0.9757 - top_k_categorical_accuracy: 0.9990
Epoch 5/5
10000/10000 [==============================] - 0s 20us/step
[7 2 1 0 4 1 4 9 5 9]
[7 2 1 0 4 1 4 9 5 9]
Accuracy: 0.98


In [177]:


# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=5,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

# Save the model to disk.
model.save_weights('model2.h5')

# Load the model from disk later using:
# model.load_weights('model.h5')


# Predict on the first 10 test images.
predictions = model.predict(test_images[:10])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) # [7, 2, 1, 0, 4]

# Check our predictions against the ground truths.
print(test_labels[:10]) # [7, 2, 1, 0, 4]

print(f'Accuracy: {accuracy(test_labels, np.argmax(predictions, axis=1))}')

Epoch 1/5
60000/60000 [==============================] - 3s 53us/step - loss: -0.8981 - top_k_categorical_accuracy: 0.9872
Epoch 2/5
60000/60000 [==============================] - 3s 58us/step - loss: -0.9432 - top_k_categorical_accuracy: 0.9959
Epoch 3/5
60000/60000 [==============================] - 3s 53us/step - loss: -0.9539 - top_k_categorical_accuracy: 0.9970
Epoch 4/5
60000/60000 [==============================] - 3s 51us/step - loss: -0.9590 - top_k_categorical_accuracy: 0.9978
Epoch 5/5
10000/10000 [==============================] - 0s 23us/step
Accuracy: 0.0


In [156]:


# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=5,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

# Save the model to disk.
model.save_weights('model3.h5')

# Load the model from disk later using:
# model.load_weights('model.h5')


# Predict on the first 10 test images.
predictions = model.predict(test_images[:10])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) # [7, 2, 1, 0, 4]

# Check our predictions against the ground truths.
print(test_labels[:10]) # [7, 2, 1, 0, 4]

print(f'Accuracy: {accuracy(test_labels, np.argmax(predictions, axis=1))}')

Epoch 1/5
60000/60000 [==============================] - 3s 58us/step - loss: -0.8977 - top_k_categorical_accuracy: 0.9869
Epoch 2/5
60000/60000 [==============================] - 3s 53us/step - loss: -0.9494 - top_k_categorical_accuracy: 0.9966
Epoch 3/5
60000/60000 [==============================] - 3s 55us/step - loss: -0.9603 - top_k_categorical_accuracy: 0.9976
Epoch 4/5
60000/60000 [==============================] - 3s 55us/step - loss: -0.9657 - top_k_categorical_accuracy: 0.9983
Epoch 5/5
10000/10000 [==============================] - 0s 21us/step


## Эксперименты и выводы
Если убрать из нормализации центровку на 0, то качество растёт на 1-2%
Попробовал заменить при этом relu на softplus - метрики упали; на tanh - метрики вернулись.

При сокращении до двух слоёв качество падает. При этом, если поставить на вход softmax, начинаются ошибки. Точость около 0.7. Увеличение слоёв не даёт значительного прироста качества.

Заменил оптимайзер на Adadelta, после чего сразу получил более 98% по всем эпохам. Вероятно, из-за того, что не перегрузил данные.
Nadam даёт 97% ; SGD - низкий результат 92% и ошибки.

Метрика top_k_categorical_accuracy дала прирост до 99% по всем эпохам. categorical_accuracy - вернула к значениям вокруг 95%
При оптимайзере Adam показатели также высокие (99%), но есть ошибки, что говорит о переобучении.

### Вывод:
Усложнение сети не даёт значимого прироста. Но и не влияет на производительность на простом примере.
Максимальный эффект дают настройки модели  
   * optimizer='Adadelta',
   * loss='cosine_similarity',
   * metrics='top_k_categorical_accuracy'


Использовались параметры из документации: 
https://keras.io/api/losses/
https://keras.io/api/optimizers/
https://keras.io/api/metrics/